### PR TITLE
Remove generator related PAT collections when running on data

### PR DESCRIPTION
**Pull request by dntaylor**
As part of a cleanup of constructed but not run modules, I am updating the runOnData portion of PAT to remove these modules. Its a simple loop over producers in the process and delattr if they are found to match the naming condition.

The full list of modules are:

packedGenParticles
patJetFlavourAssociation
patJetFlavourAssociationAK8
patJetFlavourAssociationAK8PFCHSSoftDropSubjets
patJetFlavourAssociationAK8PFPuppiSoftDropSubjets
patJetFlavourAssociationAK8Puppi
patJetFlavourAssociationLegacy
patJetFlavourAssociationLegacyAK8
patJetFlavourAssociationLegacyAK8PFCHSSoftDropSubjets
patJetFlavourAssociationLegacyAK8PFPuppiSoftDropSubjets
patJetFlavourAssociationLegacyAK8Puppi
patJetFlavourAssociationLegacyNoHF
patJetFlavourAssociationLegacyPuppi
patJetFlavourAssociationNoHF
patJetFlavourAssociationPuppi
patJetGenJetMatch
patJetGenJetMatchAK8
patJetGenJetMatchAK8PFCHSSoftDrop
patJetGenJetMatchAK8PFCHSSoftDropSubjets
patJetGenJetMatchAK8PFPuppiSoftDrop
patJetGenJetMatchAK8PFPuppiSoftDropSubjets
patJetGenJetMatchAK8Puppi
patJetGenJetMatchNoHF
patJetGenJetMatchPuppi
patJetPartonAssociationLegacy
patJetPartonAssociationLegacyAK8
patJetPartonAssociationLegacyAK8PFCHSSoftDropSubjets
patJetPartonAssociationLegacyAK8PFPuppiSoftDropSubjets
patJetPartonAssociationLegacyAK8Puppi
patJetPartonAssociationLegacyNoHF
patJetPartonAssociationLegacyPuppi
patJetPartonMatch
patJetPartonMatchAK8
patJetPartonMatchAK8PFCHSSoftDrop
patJetPartonMatchAK8PFCHSSoftDropSubjets
patJetPartonMatchAK8PFPuppiSoftDrop
patJetPartonMatchAK8PFPuppiSoftDropSubjets
patJetPartonMatchAK8Puppi
patJetPartonMatchNoHF
patJetPartonMatchPuppi
patJetPartons
patJetPartonsLegacy
patJetPartonsLegacyNoHF
patJetPartonsNoHF
prunedGenParticles
prunedGenParticlesWithStatusOne
slimmedGenJets
slimmedGenJetsAK8
tauGenJetMatch
tauGenJetMatchBoosted
tauGenJets
tauGenJetsBoosted
tauGenJetsSelectorAllHadrons
tauGenJetsSelectorAllHadronsBoosted

### CODE REVIEW

**Comment by slava77** [comment]
@cmsbuild please test

**Comment by ferencek** [comment]
I think it would be more appropriate to have all changes introduced in this PR as part of the `removeMCMatching` function.

File: PhysicsTools/PatAlgos/python/tools/coreTools.py [hunk 1]
```diff
@@ -42,6 +42,14 @@ def toolCode(self, process):
...
         print '******************* RunOnData *******************'
         removeMCMatching(process, names=names, postfix=postfix, outputModules=outputModules)
+        modStringsToRemove = [
+            'patJetGenJet', 
+            'GenParticle', 
+            'slimmedGenJets', 
+            'tauGenJet', 
+            'patJetFlavourAssociation', 
+            'patJetParton',
+        ]
```
Comment (ferencek) [inline PhysicsTools/PatAlgos/python/tools/coreTools.py]: Do you also take the postfix into account?

Comment (dntaylor) [inline PhysicsTools/PatAlgos/python/tools/coreTools.py]: It is a simple x in y check. So, yes, anything module label with one of these substrings.

**Comment by cvuosalo** [comment]
@ferencek: I think the list of modules to remove is in the right place, where Devin has put them. They should be localized to running on data.
@slava77: What do you think?

**Comment by ferencek** [comment]
@cvuosalo, I don't agree. `removeMCMatching` could be called by someone directly and it is in that function where using all the listed modules as input is being disabled. The only thing `runOnData` does explicitly is to set up the residual JECs and calling `removeMCMatching`. Therefore, `removeMCMatching` is the right place to put these updates.

**Comment by slava77** [comment]
On 8/17/16 12:28 PM, Dinko Ferencek wrote:
> @cvuosalo <https://github.com/cvuosalo>, I don't agree.
> |removeMCMatching| could be called by someone directly and it is in that
> function where using all the listed modules as input is being disabled.
> The only thing |runOnData| does explicitly is to set up the residual
> JECs and calling |removeMCMatching|. Therefore, |removeMCMatching| is
> the right place to put these updates.

Makes sense to me (all assuming this gives otherwise the same final
config/result).

**Comment by cvuosalo** [comment]
@dntaylor: Please follow if you can Dinko's suggestion about putting the list inside of removeMCMatching.

**Comment by slava77** [comment]
@cmsbuild please test

**Comment by cvuosalo** [comment]
+1

For #15472 bfe9b63201c64e9f8c614225fbfffbd5920a1a13

Removing MC-related modules that were needlessly constructed but not run for data. There should be no change in monitored quantities.

The code changes are satisfactory, and Jenkins tests against baseline CMSSW_8_1_X_2016-08-21-1100 show no significant differences, as expected. Checking the expanded configs for MC and data shows no change in the MC config and the expected removal of MC-related modules from the data config.

**Pull request merged**